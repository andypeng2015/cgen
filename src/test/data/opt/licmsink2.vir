module foo

export function w $foo(w %p1, w %p2, w %p3) {
@start:
  %n = copy.w 0_w
  %i = copy.w %p1
  jmp @loop
@loop:
  %base = add.w %p2, %p3
  %i = sub.w %i, 1_w
  %n = add.w %n, 1_w
  %c = ne.w %i, 0_w
  br %c, @loop, @exit
@exit:
  %n = copy.w 0_w
  %i = copy.w %p1
  jmp @loop2
@loop2:
  %i = sub.w %i, 1_w
  %n = add.w %n, 1_w
  %n = add.w %n, %base
  %c = ne.w %i, 0_w
  br %c, @loop2, @exit2
@exit2:
  %r = add.w %base, %n
  ret %r
}
