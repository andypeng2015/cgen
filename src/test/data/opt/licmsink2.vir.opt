module foo

export function w $foo(w %p1, w %p2, w %p3) {
@2:
  jmp @3(0x0_w, %p1)
@3(%n.2, %i.2):
  %9 = add.w %p2, %p3 ; @29
  %10 = sub.w %i.2, 0x1_w ; @30
  %12 = add.w %n.2, 0x1_w ; @32
  %11 = ne.w %10, 0x0_w ; @31
  br %11, @3(%12, %10), @11(0x0_w, %p1)
@11(%n.5, %i.5):
  %13 = sub.w %i.5, 0x1_w ; @33
  %15 = add.w %n.5, 0x1_w ; @35
  %16 = add.w %15, %9 ; @36
  %14 = ne.w %13, 0x0_w ; @34
  br %14, @11(%16, %13), @14
@14:
  %17 = add.w %9, %16 ; @37
  ret %17
}
