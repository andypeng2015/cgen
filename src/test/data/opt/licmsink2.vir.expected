module foo

export function w $foo(w %p1, w %p2, w %p3) {
@2:
  jmp @3(0x0_w, %p1)
@3(%n.2, %i.2):
  %0 = sub.w %i.2, 0x1_w ; @20
  %2 = add.w %n.2, 0x1_w ; @22
  %1 = ne.w %0, 0x0_w ; @21
  br %1, @3(%2, %0), @6
@6:
  %3 = add.w %p2, %p3 ; @23
  jmp @11(0x0_w, %p1)
@11(%n.5, %i.5):
  %4 = sub.w %i.5, 0x1_w ; @24
  %6 = add.w %n.5, 0x1_w ; @26
  %7 = add.w %6, %3 ; @27
  %5 = ne.w %4, 0x0_w ; @25
  br %5, @11(%7, %4), @14
@14:
  %8 = add.w %3, %7 ; @28
  ret %8
}
